### ISSUE TITLE
Are there any specific zip settings needed for creating an xpi?

### ISSUE THREAD

**Issue by arieljannai** [mozilla-jetpack/jpm#532]
Hi,
I know it's not exactly related, but maybe you could still have an idea.

I'm using nodejs to create an xpi with [JSZip](https://github.com/Stuk/jszip) and then I spawn the jpm process in order to sign it.
Everything looks like it's passing just fine, except that after I install it on firefox and restart it, the addon disappears.
I guess that something is wrong with the xpi creation.

I've looked in your code and saw that you also use jszip to create the xpi, and I've also tried using the same parameters as seen in your code.

Maybe some of you know if there's specific settings that needs to be applied on the archive? (Method, attributes etc..)

[I've also tried asking about it in SO](http://stackoverflow.com/questions/37509278/how-to-create-an-xpi-compatible-archive-with-nodejs)

Thanks in advance!

**Comment by Croydon** [comment]
Just wondering, if you call jpm anyway, why not generating the xpi directly with it?

**Comment by arieljannai** [comment]
It's a XUL addon, which its packing is not supported by jpm

**Comment by freaktechnik** [comment]
The signing has been outfactored into https://www.npmjs.com/package/sign-addon (not sure  jpm uses that, but that's where the fixes are afaik). Further the XPI you upload should not matter, as AMO will re-pack as far as I know. So unless you're getting back an "xpi" that's actually a JSON error it should work.

**Comment by arieljannai** [comment]
@freaktechnik Thanks, I'm trying it now

**Comment by kumar303** [comment]
> after I install it on firefox and restart it, the addon disappears

Check Tools > Web Developer > Browser Console for the real error. That should help you diagnose it.

**Comment by arieljannai** [comment]
Interesting, didn't knew this console, I was looking in the developers one.

I get:
`1464846132228	addons.xpi	WARN	Refusing to install staged add-on EXTENSION_ID with signed state -2`

I've tried looking around but couldn't find that specific error. There are other variations of the signed state (with 0 or 1), and I guess `-2` is a specific error since usually in other programs just error/fail receieve `-1` and other negative values indicates a more specific error.

Maybe some of you know that error?
Or should I ask about it in discourse/file a bugzilla?

**Comment by kumar303** [comment]
That warning is unnecessarily cryptic but I think it's linking to this: https://dxr.mozilla.org/mozilla-central/source/toolkit/mozapps/extensions/AddonManager.jsm#3100 which expands to "The downloaded file did not match the provided hash"

This is different from a signing problem. This means the file hash reported from the server did not match the real hash. We've actually seen this before in https://github.com/mozilla/addons-server/issues/1111 (although it may not be this exact issue).

What is the extension ID? If you can provide `jpm --verbose` output, that will show server headers which will be helpful.

**Comment by arieljannai** [comment]
Thanks for the help on all the platforms @kumar303 !

I've also asked a week ago on the discourse community, and I'm not sure if to continue here or there, since there's no point in double investigations in different locations.
Should we continue the discussion here or at [discourse](https://discourse.mozilla-community.org/t/signed-addon-disappear-after-installation-and-firefox-restart/8962)? (and then I'll link there to that issue)

**Comment by kumar303** [comment]
If it's a jpm bug, we should discuss it here. What is the extension ID? Can provide `jpm --verbose ...` output?

**Comment by arieljannai** [comment]
https://gist.github.com/arieljannai/2e264e8f4a2d043495cfeaf37c4012db
I haven't seen anything wrong, but maybe you'll recognize something

**Comment by kumar303** [comment]
oops, I misunderstood before. I thought you were installing the file from AMO. You are installing this file by going to File > Open in Firefox, right? `walkme_extension-0.0.204+test.dt-fx.xpi`

**Comment by arieljannai** [comment]
Yep, I just take my latest xpi and drag it into firefox, and then the popup suggests me to install it (and then restart)

**Comment by kumar303** [comment]
Could you attach the source code (before signing) so I can try to reproduce the issue?

**Comment by arieljannai** [comment]
Apperantly GitHub don't like my zip.. I've added a docx extension to the zip so it will let me upload it.
[extension.zip.docx](https://github.com/mozilla-jetpack/jpm/files/306636/extension.zip.docx)

**Comment by kumar303** [comment]
ok, thanks. Weird. I changed the ID, made an XPI, signed it, and installed it without any trouble. This was on Mac and I don't have access to a Windows machine. Maybe it's Windows related. I created the zip like `zip -r ../extension.xpi *`.

**Comment by arieljannai** [comment]
Indeed weird. That's why I thought that maybe there's something wrong with how I'm creating the zip with JSZip (though I took the example of JSZip's parameters from [jpm's code](https://github.com/mozilla-jetpack/jpm/blob/e4d58a1c1fca47aaa66ac7c187f015a2fc2a2c22/lib/xpi/utils.js#L132)).
And it's also weird cause your server is opening, reading and re-archiving the zip I'm sending. So even if it wasn't with the right parameters, it should be fine after zipping in your server, cause it's a new zip, isn't it?
